### PR TITLE
Nh auto ramp

### PR DESCRIPTION
**Pull request by noahheller**

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -223,15 +227,14 @@ private void configureBindings() {
...
+                new RampMove( ramp, () -> GameConstants.RAMP_POS_SHOOT_AMP),
                 new ShootAmp(shooter)));
 
         // Shoot the note - B
         controller.b().onTrue(CommandUtil.sequence("Operator Shoot",
                 new FeederGamepieceUntilLeave(feeder),
                 new WaitCommand(GameConstants.SHOOTER_TIME_BEFORE_STOPPING),
-                new StopShooter(shooter),
-                new RetractAmp(amp),
+                new StopShooter(shooter), new RetractAmp(amp),
```
Comment (armadilloz) [inline src/main/java/frc/robot/RobotContainer.java]: This is a no-op, right? Can you revert?

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -193,17 +195,19 @@ public void putShuffleboardCommands() {
...
+        Command alignSpeaker = CommandUtil.sequence(
                 "Shoot&AlignSpeaker",
                 new SetAlignable(drivetrain, Alignable.SPEAKER),
-                new AdvancedSpinningShot(shooter, () -> drivetrain.getPose(), () -> drivetrain.getAlignable())
+                new ParallelCommandGroup(
+                        new AdvancedSpinningShot(shooter, () -> drivetrain.getPose(), () -> drivetrain.getAlignable()),
+                        new RampFollow(ramp,() -> drivetrain.getAlignable(), () -> drivetrain.getPose())
+                )
         );
         joyLeftButton1.onTrue(alignSpeaker).onFalse(CommandUtil.logged(new SetAlignable(drivetrain, null)));
```
Comment (armadilloz) [inline src/main/java/frc/robot/RobotContainer.java:207]: Should we stop the shooter and reset the ramp on false?

File: src/main/java/frc/robot/commands/ramp/RampFollow.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                            ).getDegrees()
+            );
+            if (Constants.RAMP_DEBUG) SmartDashboard.putNumber("RAMP_TARGET_AUTO",encoderValue);
+            ramp.setRampPos(encoderValue);
+        }
+    }
+
+    @Override
+    public boolean isFinished() {
+        return alignableSupplier.get() == null;
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/ramp/RampFollow.java]: this runs the risk of not detecting the termination condition as the drivetrain may see another command set the alienable without going through null first. Better to have this be part of a race group that terminates this command, or give it a provider that tells it to finish.
And it probably needs a timeout too...

File: src/main/java/frc/robot/commands/ramp/RampFollow.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        Alignable alignableNow = alignableSupplier.get();
+        if (alignableNow != null){
+            double encoderValue = Ramp.angleToEncoder(
+                    new Rotation2d(Math.PI / 2)
+                            .minus(AutoAlignment.getYaw(alignableNow,
+                                    pose2dSupplier.get().getTranslation(),
+                                    0.5)
+                            ).getDegrees()
+            );
+            if (Constants.RAMP_DEBUG) SmartDashboard.putNumber("RAMP_TARGET_AUTO",encoderValue);
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/ramp/RampFollow.java]: "{}" around the if body.

File: src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java [hunk 5]
```diff
@@ -44,12 +39,11 @@ public void execute() {
     @Override
     public void end(boolean interrupted) {
         shooter.stopShooter();
-        timer.stop();
     }
 
     @Override
     public boolean isFinished() {
-        return timer.hasElapsed(Constants.ADVANCED_SPINNING_SHOT_TIMEOUT) || alignable.get() == null;
+        return alignable.get() == null;
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java]: Again, this runs the risk of missing the termination condition, and likely needs a timeout.

File: src/main/java/frc/robot/subsystems/Ramp.java [hunk 6]
```diff
@@ -74,8 +74,12 @@ public void resetEncoder() {
         neoPidMotor.resetEncoderPosition();
     }
 
-    public double encoderToAngle(double encoderValue){
+    public static double encoderToAngle(double encoderValue){
         //y=mx+b
         return 2.48 * encoderValue + 28.5;//needs be to measured again and put in constants
     }
+    public static double angleToEncoder(double angle){
```
Comment (armadilloz) [inline src/main/java/frc/robot/subsystems/Ramp.java:107]: why not just add setPosInDegrees()?

File: src/main/java/frc/robot/utils/AutoAlignment.java [hunk 7]
```diff
@@ -4,44 +4,90 @@
...
 
 public class AutoAlignment {
     private final static HashMap<Alignable, BiFunction<Double, Double, Rotation2d>> positionAngleMap = new HashMap<>(Map.of(
-       Alignable.AMP, (x,y)-> new Rotation2d(Math.PI/2),
-       Alignable.SPEAKER,(x,y)-> new Rotation2d(RobotContainer.isRedAlliance() ? 0 : Math.PI).plus(new Rotation2d(Math.atan(y/x))
-    )));
-    public static double calcTurnSpeed(Alignable alignable, Rotation2d currentAngle, double x, double y, PIDController controller){
-        double targetAngle = getAngle(alignable,x,y).getDegrees();
-        if (Math.abs(currentAngle.getDegrees() - targetAngle) < Constants.AUTO_ALIGN_THRESHOLD){
+            Alignable.AMP, (x, y) -> new Rotation2d(Math.PI / 2),
```
Comment (armadilloz) [inline src/main/java/frc/robot/utils/AutoAlignment.java:22]: This is too complicated to be able to follow. Needs to be simplified.

**Review by armadilloz — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (noahheller) [reply src/main/java/frc/robot/RobotContainer.java]: done (merge conflict artifact)

File: src/main/java/frc/robot/RobotContainer.java (hunk 2 above)
Comment (noahheller) [reply src/main/java/frc/robot/RobotContainer.java:207]: Shooter does stop (command returns false). I was thinking about resetting the ramp but could not decide if it was a good idea. Do you have any reasons to why we would reset it?

File: src/main/java/frc/robot/commands/ramp/RampFollow.java (hunk 3 above)
Comment (noahheller) [reply src/main/java/frc/robot/commands/ramp/RampFollow.java]: added deadline command. Regarding the timeout, I did not like the idea that they could still be holding the button and the Command times out.

File: src/main/java/frc/robot/commands/ramp/RampFollow.java (hunk 4 above)
Comment (noahheller) [reply src/main/java/frc/robot/commands/ramp/RampFollow.java]: done

File: src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java (hunk 5 above)
Comment (noahheller) [reply src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java]: added deadline command

File: src/main/java/frc/robot/subsystems/Ramp.java (hunk 6 above)
Comment (noahheller) [reply src/main/java/frc/robot/subsystems/Ramp.java:107]: great question. That makes a lot of sense

File: src/main/java/frc/robot/utils/AutoAlignment.java (hunk 7 above)
Comment (noahheller) [reply src/main/java/frc/robot/utils/AutoAlignment.java:22]: Did you have an idea in mind?

File: src/main/java/frc/robot/commands/ramp/RampFollow.java (hunk 3 above)
Comment (armadilloz) [reply src/main/java/frc/robot/commands/ramp/RampFollow.java]: In the happy scenario, they would be holding the button and the command times out, which is sad but potentially acceptable (though Bat-Chen disagrees). In the unhappy scenario the command continues to run due to a bug even after the button is released and the battery drains.

File: src/main/java/frc/robot/utils/AutoAlignment.java (hunk 7 above)
Comment (armadilloz) [reply src/main/java/frc/robot/utils/AutoAlignment.java:22]: I do. Let's talk in person once I'm in the shop.

File: src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java [hunk 8]
```diff
@@ -43,7 +43,7 @@ public void end(boolean interrupted) {
 
     @Override
     public boolean isFinished() {
-        return alignable.get() == null;
+        return false;
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java]: A comment here on why this is OK (deadline) would be nice

File: src/main/java/frc/robot/commands/ramp/RampFollow.java [hunk 9]
```diff
@@ -27,20 +27,20 @@ public RampFollow(Ramp ramp, Supplier<Alignable> alignableSupplier, Supplier<Pos
...
+                SmartDashboard.putNumber("RAMP_TARGET_Angle",targetAngle.getDegrees());
+            }
+            ramp.setAngle(targetAngle);
         }
     }
 
     @Override
     public boolean isFinished() {
-        return alignableSupplier.get() == null;
+        return false;
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/ramp/RampFollow.java]: A Comment on why this works (deadline)

File: src/main/java/frc/robot/subsystems/Ramp.java [hunk 10]
```diff
@@ -82,4 +83,8 @@ public static double angleToEncoder(double angle){
         //(y-b)/m=x
         return (angle - 28.5) / 2.48;//needs be to measured again and put in constants
     }
+
+    public void setAngle(Rotation2d angleFromGround) {
+
```
Comment (armadilloz) [inline src/main/java/frc/robot/subsystems/Ramp.java]: Something missing?

File: src/main/java/frc/robot/commands/ramp/RampFollow.java (hunk 3 above)
Comment (armadilloz) [reply src/main/java/frc/robot/commands/ramp/RampFollow.java]: Let's talk in person

File: src/main/java/frc/robot/RobotContainer.java (hunk 2 above)
Comment (armadilloz) [reply src/main/java/frc/robot/RobotContainer.java:207]: I don't see what would stop the shooter. Please inform?
Yes, probably good to reset the ramp - one less thing to do before next intake

**Review by armadilloz — APPROVED** [review]

**Pull request merged**